x-pack/plugins/lens/public/editor_frame_service/editor_frame/workspace_panel/workspace_panel.tsx

```diff
@@ -117,26 +127,60 @@ export function WorkspacePanel({
   );
 
   const [localState, setLocalState] = useState<WorkspaceState>({
-    expressionBuildError: undefined as string | undefined,
+    expressionBuildError: undefined,
     expandError: false,
   });
 
   const activeVisualization = activeVisualizationId
     ? visualizationMap[activeVisualizationId]
     : null;
+
+  const configurationValidationError = useMemo(() => {
+    const activeDatasource = activeDatasourceId ? datasourceMap[activeDatasourceId] : null;
+    const dataMessages = activeDatasourceId
+      ? activeDatasource?.getErrorMessages(datasourceStates[activeDatasourceId]?.state)
+      : undefined;
+    const vizMessages = activeVisualization?.getErrorMessages(visualizationState, framePublicAPI);
+
+    if (vizMessages || dataMessages) {
+      // Data first, visualization next
+      return [...(dataMessages || []), ...(vizMessages || [])];
+    }
+  }, [
+    activeVisualization?.getErrorMessages,
+    visualizationState,
+    activeDatasourceId,
+    datasourceMap,
+    datasourceStates,
+    framePublicAPI,
+  ]);
+
   const expression = useMemo(
     () => {
-      try {
-        return buildExpression({
-          visualization: activeVisualization,
-          visualizationState,
-          datasourceMap,
-          datasourceStates,
-          datasourceLayers: framePublicAPI.datasourceLayers,
-        });
-      } catch (e) {
-        // Most likely an error in the expression provided by a datasource or visualization
-        setLocalState((s) => ({ ...s, expressionBuildError: e.toString() }));
+      if (!configurationValidationError) {
+        try {
+          return buildExpression({
+            visualization: activeVisualization,
+            visualizationState,
+            datasourceMap,
+            datasourceStates,
+            datasourceLayers: framePublicAPI.datasourceLayers,
+          });
+        } catch (e) {
+          const buildMessages = activeVisualization?.getErrorMessages(
+            visualizationState,
+            framePublicAPI
+          );
+          const defaultMessage = {
+            shortMessage: 'One error occurred in the expression',
```

Comment:
Should be i18nified. Also, as this is shown to the user, mentioning the expression is probably confusing. What about "An unexpected error occurred while preparing the chart"?